### PR TITLE
[Merged by Bors] - sync2: do not run sync against already synced peers

### PR DESCRIPTION
**Pull request by ivan4th**
## Motivation

Initiating sync against peers which are already in sync wastes network traffic and may cause unneeded database accesses. Before initiating any sync, the node probes its peers, and after the probe it can be immediately seen which nodes are in sync with this one.

## Description

Do not sync against nodes which turn out to be in sync with this one after probing. Consider the node to be in sync and do no syncing at all if all the peers selected for sync are in sync with this node.

### CODE REVIEW

File: sync2/multipeer/multipeer.go [hunk 1]
```diff
@@ -204,11 +203,21 @@ func (mpr *MultiPeerReconciler) probePeers(ctx context.Context, syncPeers []p2p.
 	})
 
 	for pr := range probeCh {
+		if pr.InSync {
```
Comment (fasmat) [inline sync2/multipeer/multipeer.go:206]: NIT: Further up, the `for _, p := range syncPeers {` is a bit confusing to me, how about this instead. Then when the probeResult is processed and how different errors are handled is clearer to me 🙂 (also in `fullSync` it is also done like this:

```go
	for _, p := range syncPeers {
		eg.Go(func() error {
			mpr.logger.Debug("probe peer", zap.Stringer("peer", p))
			pr, err := mpr.syncBase.Probe(ctx, p)
			switch {
			case err == nil:
				probeCh <- probeResult{p, pr}
			case errors.Is(err, context.Canceled):
				return err
			default:
				mpr.logger.Warn("error probing the peer", zap.Any("peer", p), zap.Error(err))
			}
			return nil
		})
	}
```

File: sync2/rangesync/rangesync.go [hunk 2]
```diff
@@ -552,8 +555,11 @@ func (rsr *RangeSetReconciler) handleSample(
 	msg SyncMessage,
 	info RangeInfo,
 ) (pr ProbeResult, err error) {
-	pr.FP = msg.Fingerprint()
+	pr.InSync = msg.Fingerprint() == info.Fingerprint
 	pr.Count = msg.Count()
+	if pr.InSync && pr.Count != msg.Count() {
+		return ProbeResult{}, errors.New("mismatched count with matching fingerprint, collision?")
```
Comment (fasmat) [inline sync2/rangesync/rangesync.go]: NIT: errors shouldn't ask questions 😉 

```suggestion
		return ProbeResult{}, errors.New("mismatched count with matching fingerprint")
```

File: sync2/rangesync/rangesync.go [hunk 3]
```diff
@@ -552,8 +555,11 @@ func (rsr *RangeSetReconciler) handleSample(
 	msg SyncMessage,
 	info RangeInfo,
 ) (pr ProbeResult, err error) {
-	pr.FP = msg.Fingerprint()
+	pr.InSync = msg.Fingerprint() == info.Fingerprint
 	pr.Count = msg.Count()
+	if pr.InSync && pr.Count != msg.Count() {
```
Comment (fasmat) [inline sync2/rangesync/rangesync.go]: This `if` is never `true` - `pr.Count` is set to `msg.Count()` the line above, so no matter if `pr.InSync` is `true` or not the full expression will always be `false`

File: sync2/rangesync/rangesync.go [hunk 4]
```diff
@@ -552,8 +555,11 @@ func (rsr *RangeSetReconciler) handleSample(
...
+		return ProbeResult{}, errors.New("mismatched count with matching fingerprint, collision?")
+	}
 	if info.Fingerprint == msg.Fingerprint() {
```
Comment (fasmat) [inline sync2/rangesync/rangesync.go:563]: NIT: `return` instead of `else` if possible 🙂

File: sync2/multipeer/multipeer.go (hunk 1 above)
Comment (ivan4th) [reply sync2/multipeer/multipeer.go:206]: Makes sense, fixed

File: sync2/rangesync/rangesync.go (hunk 2 above)
Comment (ivan4th) [reply sync2/rangesync/rangesync.go]: Removed question but kept mention of possible collision to make a possible culprit more clear

File: sync2/rangesync/rangesync.go (hunk 3 above)
Comment (ivan4th) [reply sync2/rangesync/rangesync.go]: Fixed

File: sync2/rangesync/rangesync.go (hunk 4 above)
Comment (ivan4th) [reply sync2/rangesync/rangesync.go:563]: Fixed

**Review by fasmat — APPROVED** [review]

**Comment by ivan4th** [comment]
bors merge

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**